### PR TITLE
#788 improve Eurotronic SPZB0001 support

### PR DESCRIPTION
**Pull request by sjorge**
- [X] switch to one fromZigbee convertor (so we can do the system_mode override)
- [X] ignore hvacThermostat's system_mode 
  - [x] system_mode == heat + eurotronic_system_mode == boost => heat
  - [x] system_mode == heat + eurotronic_system_mode == !boost => auto
  - [x] system_mode == heat + eurotronic_system_mode == window open => off
- [x] rename `fromZigbee.eurotronic_system_mode` to `fromZigbee.eurotronic_host_flags`
- [x] parse eurotronic_host_flags into a hash with items per flag
- [x] add toZigbee convertors for `eurotronic_host_flags` as hash
- [x] fix eurotronic_system_mode forgets other settings

### CODE REVIEW

**Comment by allofmex** [comment]
> rename fromZigbee.eurotronic_system_mode to fromZigbee.eurotronic_host_flags

...will require adjustments in other projects too (like ioBroker). Would you also be ok to keep the name as it is? I agree that host_flags sounds closer to the real meaning, but since this is an already implemented device and users may have created scripts that would need to be changed as well...

**Comment by sjorge** [comment]
> ...will require adjustments in other projects too (like ioBroker). Would you also be ok to keep the name as it is? I agree that host_flags sounds closer to the real meaning, but since this is an already implemented device and users may have created scripts that would need to be changed as well...

I'd prefer to rename it, as the idea is to make it return a hash of the bits set (if that works out at least)

I wonder if there is a clean way to deprecate things so could keep a eurotronic_system_mode around for one zigbee2mqtt release or something giving people time to move.

Although IIRC this will cause issues with the toZigbee convertors if we have multiple ones, this still needs al of tweaking, as... this device is horribly implemented I just found out...

e.g. set system_mode=off, wait, read system_mode => heat, but if you read the host_flags window open is toggled to true! (but I digress)

**Comment by allofmex** [comment]
> I'd prefer to rename it, as the idea is to make it return a hash of the bits set

And how about keeping both, eurotronic_system_mode and additionally adding a new eurotronic_host_flags? 
```
"eurotronic_system_mode": 1, // raw bit mask of eurotronic_host_flags (like 0b00000001)
"eurotronic_host_flags": {
    "mirror_display": false,
    "boost": false,
    "window_open": false,
    "child_protection": false
  },
```
Then you have compatibility and it is only minimum overhead since the encoded value (0b00000001) has to be en/-decoded anyways

**Comment by sjorge** [comment]
@allofmex done, at least for the fromZigbee side... will do the toZigbee once I figure out how to do the host_flags hash -> bit bits and make that code also handle a 'number' as value type, then the eurotronic_system_mode should be able to use the same convertor.

**Comment by sjorge** [comment]
@allofmex @sti0 can you test these changes?

The 'system_mode' should now be correctly mapped, also there is a new 'eurotronic_host_flags' that is a object that can be used to set the bit field in an easy way. e.g.

```
mosquitto_pub -h localhost -p 1883 -t zigbee2mqtt/bedroom/radiator/set -m '{"eurotronic_host_flags": {"child_protection": true, "mirror_display": true}}'
```

Will set both the child protection and display mirroring!

There is still one bug in setting 'system_mode', I noticed it while correcting the auto vs heat behavior for system_mode... if you set system_mode, other bits for the host_flags get wiped.

This is because off gets mapped to window_open detection for example and it just sends bit 5 set.

@Koenkk I discovered a bug in the tz.eurotronic_system_mode while updating it... and I do not know how to fix it.

long story short, the real systemMode cluster is always heat... so we ignore it and use the eurotronic host flags specific cluster...

We map heat to the boost bit, off to the window open bit (TRV displays off on display).
Unfortunately this wipes the other set bits, for example child protection.

So we would need to read the current value before modifying it, but I am not sure how to do that in the convertSet code.

**Comment by sjorge** [comment]
OK, meta.state should have the data I need... will fix that too

**Comment by sti0** [comment]
@sjorge great! I'm looking forward to test it within this week. At the moment I got some trouble on my office radiator which must repaired first (not Eurotronic related ;)) I won't like to test it on my childs room ^^

**Comment by sjorge** [comment]
Alright, I fixed the setting of system_mode.... for our little device this is now entirely based on the host_flags for both reading and writing.

**Comment by sjorge** [comment]
I now get
```
zigbee2mqtt:error 2019-12-16 04:02:27: Failed to call 'EntityPublish' 'onMQTTMessage' (TypeError: Cannot read property 'key' of undefined
    at converters.find (/opt/zigbee2mqtt/lib/extension/entityPublish.js:145:56)
    at Array.find (<anonymous>)
    at EntityPublish.onMQTTMessage (/opt/zigbee2mqtt/lib/extension/entityPublish.js:145:42)
    at Controller.callExtensionMethod (/opt/zigbee2mqtt/lib/controller.js:322:44)
    at Controller.onMQTTMessage (/opt/zigbee2mqtt/lib/controller.js:237:14)
    at MQTT.emit (events.js:182:13)
    at MQTT.onMessage (/opt/zigbee2mqtt/lib/mqtt.js:86:14)
    at MqttClient.emit (events.js:182:13)
    at MqttClient._handlePublish (/opt/zigbee2mqtt/node_modules/mqtt/lib/client.js:1162:12)
    at MqttClient._handlePacket (/opt/zigbee2mqtt/node_modules/mqtt/lib/client.js:351:12))
```

When trying to set the current_heatpoint... but I am not sure why as I did not touch that code?

**Comment by sti0** [comment]
we should set the max reporting interval for the local temperature to 600sec as recommended in the docs. 
https://github.com/Koenkk/zigbee-herdsman-converters/blob/95bf9a6d79e14f84a12c34794411a34d0d987d55/devices.js#L4521

**Comment by sti0** [comment]
@sjorge I'm testing your implementation right now and switching from off->heat->boost->heat->boost->heat etc. works as expected. Even the pi_heating_demand attribute is reporting in notime now.

But i noticed an issue with the host flags.

Turning on child protection or mirroring the display by publishing 

`{"eurotronic_host_flags":{"child_protection":"true"}}`
or
`{"eurotronic_host_flags":{"mirror_display":"true"}}`
to
`zigbee2mqtt/[device]/set`
works pretty good.

But turning it off by publishing the same payload with `false` won't be recognized and it stays on `true`.

**Comment by sjorge** [comment]
> But turning it off by publishing the same payload with false won't be recognized and it stays on true.

It should be `true` and `false` without a quotes around it.

The string `"false"` will actually also evaluate to `true` in javascript/node.

 ```
[root@amethyst ~]# mosquitto_pub -h localhost -p 1883 -t zigbee2mqtt/bedroom/radiator/set -m '{"eurotronic_host_flags":{"mirror_display":true}}'
zigbee2mqtt:info  2019-12-21 11:06:24: MQTT publish: topic 'zigbee2mqtt/bedroom/radiator', payload '{"local_temperature_calibration":0,"linkquality":44,"occupied_heating_se│tpoint":16,"local_temperature":15.5,"unoccupied_heating_setpoint":16,"pi_heating_demand":8,"eurotronic_error_status":0,"current_heating_setpoint":16,"eurotronic_host_flags"│:{"mirror_display":true,"boost":false,"window_open":false,"child_protection":false},"system_mode":"auto","eurotronic_system_mode":3,"battery":100}'
[root@amethyst ~]# mosquitto_pub -h localhost -p 1883 -t zigbee2mqtt/bedroom/radiator/set -m '{"eurotronic_host_flags":{"mirror_display":false}}'
zigbee2mqtt:info  2019-12-21 11:07:25: MQTT publish: topic 'zigbee2mqtt/bedroom/radiator', payload '{"local_temperature_calibration":0,"linkquality":44,"occupied_heating_se│tpoint":16,"local_temperature":15.5,"unoccupied_heating_setpoint":16,"pi_heating_demand":8,"eurotronic_error_status":0,"current_heating_setpoint":16,"eurotronic_host_flags"│
:{"mirror_display":false,"boost":false,"window_open":false,"child_protection":false},"system_mode":"auto","eurotronic_system_mode":1,"battery":100}'
```

**Comment by sti0** [comment]
I did publish true without quotes first. Thought this didn't work. Will test it later again.

**Comment by sjorge** [comment]
@sti0 did it work 2nd try round? Would be great to get this in so I can start having a look at the reporting. (Small steps, carrying a few others diffs so want to work the diff against master away first)

**Comment by sti0** [comment]
@sjorge sorry I was busy yesterday. Tested it with true/false without quotes again and it works as expected. Tested the `system_mode` setting again and it works, too.

**Comment by sjorge** [comment]
@Koenkk this should be ready now

**Comment by sjorge** [comment]
@Koenkk anychance we can still get this in before Xmas?

**Comment by Koenkk** [comment]
Thanks @sjorge !

**Comment by sjorge** [comment]
Thanks for merging, I will try to debug some of the other reporting and get (e.g. local_temp) issues tomorrow, will be AFK for a week after tomorrow though.
